### ISSUE TITLE
Is it also possible to set the template upon sending?

### ISSUE THREAD

**Issue by satyavh** [VeliovGroup/Meteor-Mailer#2]
Cool package! But is it also possible to set the template upon sending? This would enable using multiple templates, depending on the type of message you want to send.

**Comment by dr-dimitru** [comment]
@satyavh Thank you. Why not create multiple instances with different templates?
Functionality of templates upon `send()` method was available in previous versions. It was removed with idea of main template with different content as `message` where you can use any king of HTML code.
But I'm open for this request if you can describe use cases of this feature.

**Comment by satyavh** [comment]
Yes, that's an option. But I see it's very easy to implement (and there's even reference in the code still - line 90) upon send so I might just do that.

Basically, I just want to do this:

```
Meteor.mail.send 'to@example.com',
 message: "Some HTML or plain-text string (required)"
 subject: "Some HTML or plain-text string (required)"
 template: 'emailTemplates/signUp.html'
 #Any optional keys, like
```

and upon send the template that you specify is rendered (with the optional keys), instead of the global template. 

The use case is that for each type of email message you might want to dynamically insert different data. It can all be stuffed in the {{message}} but it's more powerful to setup different helpers for each email. Otherwise you need to construct hard to follow message html like for example:

```
      message: "You've been made an " + invite.role + " by " + name + '. <a href="' + baseUrl + 'register/' + invite.token + '" target="_blank" >Click here to setup your account and get started.</a>'
```

Which should basically be in the template like this

```
<p>You've been made a {{role}} by {{name}}. <a href="{{baseUrl}}register/{{token}}" target="_blank">Click here to setup your account and get started.</a></p>
```

And then you pass role, baseUrl, token to the template. But another email message will probably required different helpers.

And that means that you need different templates for each message.

**Comment by dr-dimitru** [comment]
Agree, but what are you saying ,means we are need to implement Layout support which can be defined in instance creation (aside to default template), then use template inside `send()` method, right? If you agree - I can implement this in few days

**Comment by dr-dimitru** [comment]
Added at `v1.0.1` please update, and let me know how it work for your end

**Comment by satyavh** [comment]
Hmm I can't make it work. My settings are:

```
  Meteor.mail = new Meteor.Mailer
    login: 'xxx'
    host: 'xxx'
    connectionUrl: 'xxx'
    accountName: 'Pipetracker'
```

and then upon send:

```
    Meteor.mail.send invite.name + '<' + invite.email + '>',
      subject: name + " invited you to join Pipetracker."
      inviter: name
      role: invite.role
      inviteUrl: baseUrl + 'register/' + invite.token
      btnText: "Get started"
    ,
      template: 'emailTemplates/email.html'
```

The template is located in `/private/emailTemplates`, it should be relative to private, right?
But it keeps sending the default email template. 

The problem might be that it only reads template setting if you pass in sendAt and a callback?

**Comment by dr-dimitru** [comment]
Try like this:

``` coffeescript
Meteor.mail.send "#{invite.name}<#{invite.email}>",
   subject: name + " invited you to join Pipetracker."
   inviter: name
   role: invite.role
   inviteUrl: baseUrl + 'register/' + invite.token
   btnText: "Get started"
,
   null
,
   null
,
   template: 'emailTemplates/email.html'
```

Yes, asset should be relative to ~~public~~ `private` folder.

**Comment by satyavh** [comment]
Passing null, null: it seems to get stuck in the mail queue. 

Are you sure it should be relative to the public folder? I read in the Meteor docs that assets should be relative to the private folder.

> Assets allows server code in a Meteor application to access static server assets, which are located in the private subdirectory of an application's tree.

I think a better / more clear way to structure this would be passing in an object:

```
Meteor.mail.send 
  recipient: 'email@example.com'
  subject: 'whatever'
  helpers: 
     helper1: 'hello'
     helper2: 'hi'
     ...etc
  sendAt: new Date()
  callback:   (error, success, recipient) ->
    console.log("mail is not sent to #{recipient}") if error
    console.log("mail successfully sent to #{recipient}") if success
  template: '/templates/mytemplate.html'
```

that way you can explicitly check for keys to exists, and you don't have to pass in `null` values or keep in mind the order of the arguments.

**Comment by dr-dimitru** [comment]
You are right - I event thought to write `private`, but somewhy did a mistake, sorry for this.
Agree with you about object - will done it un further update.
Let me know of current version works for you

**Comment by satyavh** [comment]
I was working on a PR to implement this, but then I found out that the way you try to load templates with this package is **not going to work**. You can't use `Assets.getText()` from a package, so unfortunately this renders your package a little useless.

There are workarounds, like Lookback has done, read here:
https://github.com/lookback/meteor-emails/issues/4

I've solved this in my PR by providing the template like this:
`template: Assets.getText 'mytemplate.html'` to the Meteor.mail.send call.

But whether you accept my PR or not, you'll have to update this to make this package work :-)

**Comment by satyavh** [comment]
Another thing that's not working: the callback. You can't just store a .js function in a Mongo collection, it will be dropped and not saved. Read more here

https://docs.mongodb.org/v3.0/tutorial/store-javascript-function-on-server/

**Comment by dr-dimitru** [comment]
Please see [v1.1.0](https://github.com/VeliovGroup/Meteor-Mailer/releases/tag/v1.1.0)
Has minor changes from your PR, callbacks is fixed, thank you for contribution
Please let me know how it goes on your end

**Comment by dr-dimitru** [comment]
@satyavh Please, let me know how latest update works on your end

**Comment by satyavh** [comment]
Yes I will do that this week. I'm very busy but I promise.

**Comment by dr-dimitru** [comment]
Thank you on this :sunglasses:

**Comment by satyavh** [comment]
Hi, it works like a charm. Helpers, callbacks, looking very good!

**Comment by dr-dimitru** [comment]
@satyavh  Thank you for contribution

**Comment by sean-stanley** [comment]
Could the README be updated to reflect new use of helpers and such?

Looks like a nice light-weight package. Good job.

**Comment by dr-dimitru** [comment]
Hi @sean-stanley ,
There is no actual helpers in general meaning as in classic Meteor Templates.
It simply pick ups any of properties passed in `options` (first argument of `send()` method). And replaces placeholders defined in double curly brackets, like `{{placeholder}}` and `MailInstance.send({placeholder: 'value'})`

Hope this little extra info helps. Let me know please.

**Comment by sean-stanley** [comment]
So which of these would work?

```
Meteor.mail.send
... 
  appname: ->
    Collection.findOne()
  user: ->
    Meteor.user()
  stringArgumentFromMethod: StringArgumentFromMethod
```

**Comment by dr-dimitru** [comment]
Nope, not the functions, only if you will write it this way:

``` coffeescript
Meteor.mail.send
... 
  appname: do ->
    Collection.findOne()
  user: do ->
    Meteor.user()
  stringArgumentFromMethod: StringArgumentFromMethod
```
